File: services/api/packages/api/src/sync_api/messaging/placeholders.py
from __future__ import annotations

import re
from dataclasses import dataclass, fields
from typing import TYPE_CHECKING, Final

if TYPE_CHECKING:
    from collections.abc import Iterable

_PLACEHOLDER: Final = re.compile(r"\{\{(?P<name>[^{}]*)\}\}")


@dataclass(frozen=True, slots=True)
class Placeholders:
    """Everything a Message template may name, and what one send fills each with.

    Declaring the set once as these fields is what keeps `KNOWN` — which save time validates
    against — from ever admitting a name send time could not resolve.
    """

    candidate_name: str
    job_title: str
    tenant_name: str

    def fill(self, template: str) -> str:
        values = {field.name: getattr(self, field.name) for field in fields(self)}
        return _PLACEHOLDER.sub(lambda found: values[_named(found)], template)


KNOWN: Final = tuple(field.name for field in fields(Placeholders))


def as_written(names: Iterable[str]) -> str:
    """The names as a recruiter writes them, for saying which ones a template may use."""
    return ", ".join(f"`{{{{ {name} }}}}`" for name in names)


def unknown_in(template: str) -> list[str]:
    """Every placeholder the text names that a send could not fill, in the order written.

    A malformed one lands here too: `{{ Job Title }}` names nothing resolvable, and saying so at
    save time is the whole point of refusing it there.
    """
    named = (_named(found) for found in _PLACEHOLDER.finditer(template))
    return list(dict.fromkeys(name for name in named if name not in KNOWN))


def _named(found: re.Match[str]) -> str:
    """Whitespace inside the braces is the recruiter's, so `{{name}}` and `{{ name }}` are one."""
    return found.group("name").strip()

File: apps/recruiter-portal/src/features/templates/placeholders.ts
export const FILLABLE: readonly string[] = ['candidate_name', 'job_title', 'tenant_name'];

export const PLACEHOLDER = /\{\{([^{}]*)\}\}/g;

function asWritten(name: string): string {
  return `{{ ${name} }}`;
}

export function asList(names: readonly string[]): string {
  const written = names.map(asWritten);
  if (written.length < 2) return written.join('');
  return `${written.slice(0, -1).join(', ')} or ${written.at(-1)}`;
}

function unfillableIn(text: string): string[] {
  const named = [...text.matchAll(PLACEHOLDER)].map((found) => (found[1] ?? '').trim());
  return [...new Set(named.filter((name) => !FILLABLE.includes(name)))];
}

export function unfillableMessage(text: string): string | null {
  const unfillable = unfillableIn(text);
  if (unfillable.length === 0) return null;
  return `Nothing can fill ${asList(unfillable)}. Use ${asList(FILLABLE)}.`;
}
